### PR TITLE
Add Cloud Build for GCP CI env

### PR DESCRIPTION
**Pull request by roger2hk**
Towards #53

### CODE REVIEW

File: deployment/modules/gcp/cloudbuild/variables.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+variable "env" {
+  description = "Unique identifier for the env, e.g. dev or ci or prod"
+  type        = string
+}
+
+variable "docker_env" {
+  description = "Unique identifier for the Docker env, e.g. dev or ci or prod"
+  type        = string
+}
```
[the remark this answers is not in the document]
Comment (roger2hk) [reply deployment/modules/gcp/cloudbuild/variables.tf]: Done.

File: deployment/live/gcp/cloudbuild/terragrunt.hcl [hunk 2]
```diff
@@ -0,0 +1,22 @@
+locals {
+  env        = path_relative_to_include()
+  project_id = get_env("GOOGLE_PROJECT", "transparency-dev")
```
Comment (phbnf) [inline deployment/live/gcp/cloudbuild/terragrunt.hcl:3]: Did you mean to put static-ct?

File: deployment/modules/gcp/cloudbuild/main.tf [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  member  = "serviceAccount:${google_service_account.cloudbuild_service_account.email}"
+}
+
+resource "google_cloudbuild_trigger" "build_trigger" {
+  name            = "build-docker-${var.docker_env}"
+  service_account = google_service_account.cloudbuild_service_account.id
+  location        = var.location
+
+  github {
+    owner = "transparency-dev"
```
Comment (phbnf) [inline deployment/modules/gcp/cloudbuild/main.tf:79]: Maaaybe make this a parameter such that people can use this with a repo fork if they want? The tessera repo doesn't do this though, I just did it myself to try this out on my personal project.

File: deployment/live/gcp/cloudbuild/terragrunt.hcl [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  location   = get_env("GOOGLE_REGION", "us-central1")
```
Comment (phbnf) [inline deployment/live/gcp/cloudbuild/terragrunt.hcl:4]: very nit, region to align with tessera repo

File: deployment/live/gcp/cloudbuild/prod/terragrunt.hcl [hunk 5]
```diff
@@ -0,0 +1,18 @@
+terraform {
+  source = "${get_repo_root()}/deployment/modules/gcp//cloudbuild"
+}
```
Comment (phbnf) [inline deployment/live/gcp/cloudbuild/prod/terragrunt.hcl:3]: can this be moved to the ../terragrunt.hcl file?

File: deployment/live/gcp/cloudbuild/prod/terragrunt.hcl [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+locals {
+  docker_env = "ci"
+  server_docker_image = "${include.root.locals.location}-docker.pkg.dev/${include.root.locals.project_id}/docker-${local.docker_env}/conformance-gcp:latest"
```
Comment (phbnf) [inline deployment/live/gcp/cloudbuild/prod/terragrunt.hcl:7]: Is this required? I don't think it's being used.

File: deployment/live/gcp/cloudbuild/prod/terragrunt.hcl (hunk 5 above)
Comment (roger2hk) [reply deployment/live/gcp/cloudbuild/prod/terragrunt.hcl]: Moved.

File: deployment/live/gcp/cloudbuild/prod/terragrunt.hcl (hunk 6 above)
Comment (roger2hk) [reply deployment/live/gcp/cloudbuild/prod/terragrunt.hcl]: Removed.

File: deployment/live/gcp/cloudbuild/terragrunt.hcl (hunk 2 above)
Comment (roger2hk) [reply deployment/live/gcp/cloudbuild/terragrunt.hcl]: Updated to `static-ct`.

File: deployment/live/gcp/cloudbuild/terragrunt.hcl (hunk 4 above)
Comment (roger2hk) [reply deployment/live/gcp/cloudbuild/terragrunt.hcl:8]: Do you mean the `location` cannot be changed?

File: deployment/modules/gcp/cloudbuild/main.tf (hunk 3 above)
Comment (roger2hk) [reply deployment/modules/gcp/cloudbuild/main.tf]: Done.

File: deployment/modules/gcp/cloudbuild/main.tf [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  project = var.project_id
+  role    = "roles/run.developer"
+  member  = "serviceAccount:${google_service_account.cloudbuild_service_account.email}"
+}
+
+resource "google_project_iam_member" "iam_service_account_user" {
+  project = var.project_id
+  role    = "roles/iam.serviceAccountUser"
+  member  = "serviceAccount:${google_service_account.cloudbuild_service_account.email}"
+}
```
Comment (phbnf) [inline deployment/modules/gcp/cloudbuild/main.tf:71]: I think this all defined in overground, no need to define this here.

File: deployment/modules/gcp/cloudrun/main.tf [hunk 8]
```diff
@@ -19,6 +19,12 @@ resource "google_service_account" "cloudrun_service_account" {
   display_name = "Service Account for Cloud Run (${var.env})"
 }
 
+resource "google_project_iam_member" "run_service_agent" {
+  project = var.project_id
+  role    = "roles/run.serviceAgent"
+  member  = "serviceAccount:${google_service_account.cloudrun_service_account.email}"
+}
+
```
Comment (phbnf) [inline deployment/modules/gcp/cloudrun/main.tf:27]: Same here, and for the other membership below actually (I had missed this in a previous PR). I don't think they should live here, but in Overground rather. Then, terraform can add biding to the GCS and Spanner resources for the right users to have the right roles on these resources. Let's clean that all up in a followup PR?

Or one could argue that we want to leave them around for transparency and to show to the world what's our IAM structure, at the cost of having two sources for this... TBD speaking with the team I'd say.

**Review by phbnf — APPROVED** [review]

File: deployment/modules/gcp/cloudbuild/main.tf (hunk 7 above)
Comment (roger2hk) [reply deployment/modules/gcp/cloudbuild/main.tf:71]: They're useful when setting up the cloud build in other GCP projects. I'm inclined to keep these. Perhaps we can use conditional IAM resources creation or provide the `terraform import` command.

File: deployment/live/gcp/cloudbuild/terragrunt.hcl (hunk 4 above)
Comment (roger2hk) [reply deployment/live/gcp/cloudbuild/terragrunt.hcl]: I see. They seem to be interchangeable. Most terraform resources use `location` as the arguments.

File: deployment/modules/gcp/cloudrun/main.tf (hunk 8 above)
Comment (roger2hk) [reply deployment/modules/gcp/cloudrun/main.tf:27]: Sure, let's talk about this with the team.

**Pull request merged**